### PR TITLE
Make sure components in nested live views do not conflict

### PR DESCRIPTION
**Pull request by josevalim**
Imagine this setup:

    LiveView
      LiveView [id=child_live]
        SomeComponent, id: :stateful
      SomeComponent, id: :also_stateful

On the client, both components will have the same
data-phx-component=ID. Previously, the JS client was
running selectors on the whole document, which meant
clicking on the component "also_stateful" would change
the "stateful" one.

This PR fixes it by doing two changes:

  1. We moved the withinTargets function to inside the View object.
  This makes sure we always scope the query selectors properly

  2. We changed findComponentNodeList to remove all components that
  are children of a child LiveView

### CODE REVIEW

File: assets/js/phoenix_live_view.js [hunk 1]
```diff
@@ -1568,6 +1563,21 @@ export class View {
...
+    if(/^(0|[1-9](\d?)+)$/.test(phxTarget)){
+      let targets = DOM.findComponentNodeList(this.el, phxTarget)
+      if(targets.length === 0){
+        logError(`no phx-target's found matching @myself of ${phxTarget}`)
+      } else {
+        callback(targets[0])
+      }
+    } else {
+      let targets = Array.from(this.el.querySelectorAll(phxTarget))
+      if(targets.length === 0){ logError(`no phx-target's found for selector "${phxTarget}"`) }
```
Comment (josevalim) [inline assets/js/phoenix_live_view.js:1576]: I changed this to `logError` because race conditions may actually mean that nothing will be found.

**Comment by henrik** [comment]
Also see #995, now closed because it would conflict with this one. Suggested new error messages:

```
if(!myselfTarget){ throw new Error(`nothing found matching the phx-target @myself value ${phxTarget}`) }

if(targets.length === 0){ throw new Error(`nothing found matching the phx-target selector "${phxTarget}"`) }
```

Not sure if #963 would be suitable to fold into this one as well, since it's the same code block? It suggests simplifying the regex like so:

``` diff
-    if(/^(0|[1-9](\d?)+)$/.test(phxTarget)){
+    if(/^(0|[1-9]\d*)$/.test(phxTarget)){
```

File: assets/js/phoenix_live_view.js [hunk 2]
```diff
@@ -1593,6 +1588,21 @@ export class View {
...
+    if(/^(0|[1-9](\d?)+)$/.test(phxTarget)){
+      let targets = DOM.findComponentNodeList(this.el, phxTarget)
+      if(targets.length === 0){
+        logError(`no phx-target's found matching @myself of ${phxTarget}`)
+      } else {
+        callback(this, targets[0])
+      }
+    } else {
+      let targets = Array.from(document.querySelectorAll(phxTarget))
+      if(targets.length === 0){ logError(`no phx-target's found for selector "${phxTarget}"`) }
```
Comment (josevalim) [inline assets/js/phoenix_live_view.js:1601]:
```suggestion
      if(targets.length === 0){ logError(`nothing found matching the phx-target selector "${phxTarget}"`) }
```

File: assets/js/phoenix_live_view.js [hunk 3]
```diff
@@ -1593,6 +1588,21 @@ export class View {
     this.liveSocket.log(this, kind, msgCallback)
   }
 
+  withinTargets(phxTarget, callback){
+    if(/^(0|[1-9](\d?)+)$/.test(phxTarget)){
+      let targets = DOM.findComponentNodeList(this.el, phxTarget)
+      if(targets.length === 0){
+        logError(`no phx-target's found matching @myself of ${phxTarget}`)
```
Comment (josevalim) [inline assets/js/phoenix_live_view.js:1595]:
```suggestion
        logError(`no component found matching phx-target of ${phxTarget}`)
```

File: assets/js/phoenix_live_view.js [hunk 4]
```diff
@@ -1007,9 +995,16 @@ export let DOM = {
     return callback ? array.forEach(callback) : array
   },
 
-  findFirstComponentNode(node, cid){ return node.querySelector(`[${PHX_COMPONENT}="${cid}"]`) },
+  findComponentNodeList(node, cid){
+    let phxChildren = this.all(node, `${PHX_VIEW_SELECTOR}`)
```
Comment (henrik) [inline assets/js/phoenix_live_view.js:999]:
```suggestion
    let phxChildren = this.all(node, PHX_VIEW_SELECTOR)
```

File: assets/js/phoenix_live_view.js [hunk 5]
```diff
@@ -1007,9 +995,16 @@ export let DOM = {
     return callback ? array.forEach(callback) : array
   },
 
-  findFirstComponentNode(node, cid){ return node.querySelector(`[${PHX_COMPONENT}="${cid}"]`) },
+  findComponentNodeList(node, cid){
+    let phxChildren = this.all(node, PHX_VIEW_SELECTOR)
+    let result = this.all(node, `[${PHX_COMPONENT}="${cid}"]`)
 
-  findComponentNodeList(node, cid){ return this.all(node, `[${PHX_COMPONENT}="${cid}"]`) },
+    if(phxChildren.length === 0) {
```
Comment (chrismccord) [inline assets/js/phoenix_live_view.js:1002]:
```suggestion
    if(phxChildren.length === 0){
```

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**